Packet++/src/Packet.cpp

```diff
@@ -57,14 +57,17 @@ void Packet::setRawPacket(RawPacket* rawPacket, bool freeRawPacket, ProtocolType
 
 	m_LastLayer = m_FirstLayer;
 	Layer* curLayer = m_FirstLayer;
-	while (curLayer != NULL && (curLayer->getProtocol() & parseUntil) == 0 && curLayer->getOsiModelLayer() <= parseUntilLayer)
+	int counter = 0;
+	while (curLayer != NULL && (curLayer->getProtocol() & parseUntil) == 0 && curLayer->getOsiModelLayer() <= parseUntilLayer && counter < 64)
 	{
 		m_ProtocolTypes |= curLayer->getProtocol();
 		curLayer->parseNextLayer();
 		curLayer->m_IsAllocatedInPacket = true;
 		curLayer = curLayer->getNextLayer();
+		
 		if (curLayer != NULL)
 			m_LastLayer = curLayer;
+		counter++;

```

Comment:
I didn't have time to get into the packet example you provided. Can you explain what causes the infinite loop?

If it's a specific protocol maybe we can fix there...

Reply:
The underlying problem seems to be here: https://github.com/seladb/PcapPlusPlus/blob/f30de1387acc1f823e0d4779e5565f4cc32dd4a6/Packet%2B%2B/src/IPv4Layer.cpp#L240

For the provided packet example, `hdrLen` seems to be 0, so that the payload of the new packet, is at the same position as the payload of the current packet, when calculating the position of the payload here: https://github.com/seladb/PcapPlusPlus/blob/f0f5a613411798d75804ae7815c847a405c6f460/Packet%2B%2B/src/IPv4Layer.cpp#L251

I guess that means, that the exact same data is parsed all the time when `hdrLen` is 0, which results in an infinite loop. I reverted the bug fix from above and commited a new one, which checks whether `hdrLen` is not 0. I hope that fix is okay, as a header with length 0 is malformed anyway, right? Also, this bug might also occur for other layers, if they also assume that `hdrLen` is always valid.

Reply:
I also added the example crash files to the folder containing the regression tests